### PR TITLE
Fix issue #4010

### PR DESCRIPTION
**Pull request by BachiLi**
In autodiff, legalize adjoint function naming, and don't propagate to non float parameters.

### CODE REVIEW

File: test/correctness/autodiff.cpp [hunk 1]
```diff
@@ -1246,7 +1246,8 @@ void test_select_guard() {
 }
 
 void test_param() {
-    Param<float> param(2.f);
+    // Give the parameter an illegal name to test the function renaming.
+    Param<float> param("param.0", 2.f);
```
Comment (abadams) [inline test/correctness/autodiff.cpp]: We may forbid explicit param names with dots at some point. Instead, make it a buffer param, and use its width where you use param below.

Comment (BachiLi) [reply test/correctness/autodiff.cpp]: Like this? https://github.com/halide/Halide/pull/4016/commits/bac5e9be0e113b65cecce2cb4a01d02d5223356b

Comment (BachiLi) [reply test/correctness/autodiff.cpp]: Accidentally turned out all other tests. Fixed in https://github.com/halide/Halide/pull/4016/commits/1e4076d653b7bc5459b99a3205efde923d0acf9d

Comment (BachiLi) [reply test/correctness/autodiff.cpp]: Sorry, my brain wasn't working. Fixed the test by using an ImageParam. See https://github.com/halide/Halide/pull/4016/commits/e7ca5649a079d66ab72a24af582cfe76ed3aaa7c

**Comment by steven-johnson** [comment]
Is this ready to land?

**Comment by BachiLi** [comment]
I think it is ready but @abadams should take a quick look first.

**Pull request merged**